Move the HvAUX struct into the end of the hash's body

nwc10 opened the pull request:
Memory for hash iterators *etc* is only allocated on demand, as most hashes are used as objects and so are never iterated.

The iterator state (and other extra data) is stored in a `struct xpvhv_aux`. In blead this is allocated as part of the block of memory used to hold the linked lists of `HE *` pointers - the hash's main array. As a consequence, the address of this struct can move if the hash grows, which adds complexity in quite a few places.

Instead, allocate the memory as part of the hash body - this way, the memory doesn't move once it is allocated.

Memory is still only allocated on demand - "upgrade" the body as needed, changing the existing code paths that (re)allocated `HvARRAY()` to instead swap the body.

Larger bodies can still be allocated from arenas - we can use the arena for `SVt_IV`. At one point this arena was used to allocate memory for pointer table hashes, until that allocation was moved into the pointer table code.

ASAN does not find this branch interesting (no regressions, threaded or unthreaded, with `-DPURIFY`)

nwc10 committed:
Inline the xhv_aux struct in the main hash body

For now, memory for this structure is always allocated, even though it
isn't always flagged as being present.

nwc10 committed:
Delete the do_aux code from S_hsplit()

This existed to handle automatically allocating an "aux" structure for
larger hashes, to save a (large) reallocation if they needed to be iterated.

Now that we no longer store the aux structure in the main hash array, we
don't need to take this precaution.

nwc10 committed:
Inline S_hv_auxinit_internal() into S_hv_auxinit()

The previous commit eliminated the only other use of S_hv_auxinit_internal().

nwc10 committed:
Change S_new_body to static inline, and use it instead of a macro

Rename the macro new_body_inline() to new_body_from_arena().

The macro existed to provide an inline implementation for hot code paths,
with a small function for the others. Now that we have proper inline
functions, change the function to be inlined, and use it instead of the
macro. There is no change to the code generated for the hot paths. This
makes the source simpler.

Yes, this will also inline the function in a couple of other places, but
these days optimising compilers often do that anyway for small functions.
Even if they aren't for this function (as was), the size of the function
itself is comparable with the calling overheads, so it's not a massive
penalty.

Two if blocks now need to be placed inside #ifndef PURIFY/#endif guards
because they now use the function S_new_body(), which is not defined with
PURIFY. The code is actually unreachable - defining PURIFY causes all
values for new_type_details->arena to be forced to 0 - hence the if was
never true under PURIFY. Hence it could have been guarded the same way
previously, but for (maybe false) improved readability was not. Previously
there wasn't a compilation failure because the unreachable code was using
a macro, not a static function, and the macro was always defined, even when
not used.

Really the macro's definition should have been inside #ifndef PURIFY, so
fix this oversight - it is now only defined if it is needed.

Retain the macro (with a new name), and split the parameters to the macro.
This will make the next commit simpler.

nwc10 committed:
Split the XPVHV body into two variants "normal" and "with aux"

Default to the smaller body, and switch to the larger body if we need to
allocate a C<struct xpvhv_aux> (eg need an iterator).

This restores the previous small size optimisation for hashes used as
objects.

nwc10 committed:
Re-reading HvAUX(hv) is no longer necessary

Now that HvAUX(hv) is allocated as part of the HV's body, it no longer
moves in memory if the hash is resized. Hence there is no need to update
local copies of the pointer after any potential move, so remove the code
that did this.

hvds commented on hv.c:
```diff
@@ -1460,30 +1451,6 @@ S_hsplit(pTHX_ HV *hv, STRLEN const oldsize, STRLEN newsize)
-        struct xpvhv_aux *const dest = HvAUX(hv);
-        if (SvOOK(hv)) {
-            /* we reset the iterator's xhv_rand as well, so they get a totally new ordering */
-#ifdef PERL_HASH_RANDOMIZE_KEYS
-            dest->xhv_rand = (U32)PL_hash_rand_bits;
-#endif
```
If I read things correctly, this was already a noop (since we've already set it on insert before deciding we need a split). If that's correct, it might be worth removing these 3 lines as a separate commit; if it's not, it probably needs mentioning in the commit message that we now no longer re-randomize on split.

nwc10 replied:
> If I read things correctly

Oh my. Yes, I've been going backwards and forwards for a while trying to be sure about this.

I think that you are correct that it's a no-op for an insert. (Well spotted. I'd not even noticed that this might be relevant.)

But. It doesn't seem to be a no-op where `S_hsplit` is called by `Perl_hv_ksplit`, which is `keys` as an LVALUE where the hash is expanded. So things like `perl -MDevel::Peek -lwe '%h = (1..4); Dump \%h; print scalar each %h; Dump \%h; keys %h = 20; Dump \%h; print scalar each %h; Dump \%h'`

In blead that resets the iterator *and* forces a new random ordering. In this branch, as a result of the commit above it doesn't. (The line `RAND = 0x......` in the rather verbose debug output)

I suspect that the correct thing is to split those 3 lines out as a separate commit that also inserts them into `Perl_hv_ksplit`, along with comments explaining what's going on, but I want to sleep on this idea.

hvds replied:
> It doesn't seem to be a no-op where `S_hsplit` is called by `Perl_hv_ksplit`

Ah right, I hadn't considered lvalue keys().

> I suspect that the correct thing is to split those 3 lines out as a separate commit that also inserts them into `Perl_hv_ksplit`, along with comments explaining what's going on, but I want to sleep on this idea.

That makes sense, if seed reset is really required for this case. It isn't obvious to me that it _is_ required (other than for continuity with current behaviour) for a request to change the number of buckets that doesn't change the actual list of keys.

nwc10 replied:
> That makes sense, if seed reset is really required for this case. It isn't obvious to me that it _is_ required (other than for continuity with current behaviour) for a request to change the number of buckets that doesn't change the actual list of keys.

Agree with your reasoning. But it was easy enough to keep the current behaviour, so (for now) I did, as e9dd222847f9

hvds commented on sv.c:
```diff
@@ -1417,12 +1414,15 @@ Perl_sv_upgrade(pTHX_ SV *const sv, svtype new_type)
         assert(new_type_details->body_size);
         /* We always allocated the full length item with PURIFY. To do this
            we fake things so that arena is false for all 16 types..  */
+#ifndef PURIFY
```
It isn't clear to me from the commit message why this block and the one below newly become not-PURIFY only. Is that an additional bugfix, an additional optimization, or is it somehow newly required by the rest of this change?

nwc10 replied:
> It isn't clear to me from the commit message why this block and the one below newly become not-PURIFY only. Is that an additional bugfix, an additional optimization, or is it somehow newly required by the rest of this change?

It wasn't actually obvious to me either, coming back to this. (The branch is basically a group of cherry-picks of commits from early August). "newly required" is the answer, now documented in the commit message to 6d91a202e2c91c7b. As a side effect of you asking this, I spotted some more things I thought could be cleared up better, so they are also now in that commit.

hvds replied:
> "newly required" is the answer, now documented in the commit message to [6d91a20](https://github.com/Perl/perl5/commit/6d91a202e2c91c7bcafe475970b333db8863e7d7).

Cool, makes much more sense now thanks. In that commit message, though, `s/as/was/` in "the unreachable code as using".

nwc10 replied:
Well spotted. Fixed in the rebase that I just pushed

hvds commented on sv.c:
```diff
@@ -1496,6 +1507,42 @@ Perl_sv_upgrade(pTHX_ SV *const sv, svtype new_type)
+#ifdef PURIFY
+    new_body = new_NOARENAZ(&fake_hv_with_aux);
+#else
+    new_body_inline(new_body, SVt_IV, fake_hv_with_aux);
```
I think this could do with a comment to explain the `SVt_IV`. I suspect it somehow links with the renaming of that parameter from `svtype` to `root_index` (in the macro, but not the function), but I don't really have a clue. (And I'm not sure `new_body_inline` is still the most appropriate name for the macro.)

nwc10 replied:
> I think this could do with a comment to explain the `SVt_IV`. I suspect it somehow links with the renaming of that parameter from `svtype` to `root_index` (in the macro, but not the function), but I don't really have a clue. (And I'm not sure `new_body_inline` is still the most appropriate name for the macro.)

`SVt_IV` became `#  define HVAUX_ARENA_ROOT_IX   SVt_IV` in `sv.h` and more comments.

Yes, `new_body_inline` wasn't really the best name. It's now `new_body_from_arena` - if there's a better name still, it eludes me.

I also found some outdated comments in `sv.c` to eliminate.

nwc10 committed:
No need to set dest->xhv_rand in S_hsplit() called from hv_common()

Perl_hv_common() only calls S_hsplit() after inserting a new hash key, if
the hash meets the "split" criteria. Inserting the new hash key has already
set HvAUX(hv)->xhv_rand to PL_hash_rand_bits (if the hash has an aux
struct), so no need to repeat the same assignment in S_hsplit() for this
case.

However, this code path in S_hsplit() can also be reached from
Perl_hv_ksplit() for the case of assigning to `keys %hash` for a hash which
has an aux struct. The call to `keys` resets the hash iterator (in LVALUE
context as well as RVALUE context), but when the assignment changed the
bucket size this also had the size effect of causing a new iteration order.

It's not clear whether this really matters for security of the internal hash
state, but as this behaviour is both observable and can easily be retained,
it seems worthwhile adding the small amount of extra code needed to keep it.

Hence this commit is (or is intended to be) a pure refactoring, without any
observable behaviour change.

nwc10 committed:
Rename HE_SVSLOT to HE_ARENA_ROOT_IX

The longer name more accurately reflects what the constant refers to.

Correct the comments describing how some arena roots are re-used.

nwc10 committed:
Move variables in Perl_sv_clear() to a tighter scope.

These three variables don't hold their values between iterations of the
while loop, so can be declared within it, to make this obvious.

demerphq commented:
On Sun, 26 Sept 2021 at 20:58, Nicholas Clark ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In hv.c <https://github.com/Perl/perl5/pull/19146#discussion_r716246111>:
>
> > -#ifdef PERL_HASH_RANDOMIZE_KEYS
> -            dest->xhv_rand = (U32)PL_hash_rand_bits;
> -#endif
>
> If I read things correctly
>
> Oh my. Yes, I've been going backwards and forwards for a while trying to
> be sure about this.
>
> I think that you are correct that it's a no-op for an insert. (Well
> spotted. I'd not even noticed that this might be relevant.)
>
> But. It doesn't seem to be a no-op where S_hsplit is called by
> Perl_hv_ksplit, which is keys as an LVALUE where the hash is expanded. So
> things like perl -MDevel::Peek -lwe '%h = (1..4); Dump \%h; print scalar
> each %h; Dump \%h; keys %h = 20; Dump \%h; print scalar each %h; Dump \%h'
>
> In blead that resets the iterator *and* forces a new random ordering. In
> this branch, as a result of the commit above it doesn't. (The line RAND =
> 0x...... in the rather verbose debug output)
>
> I suspect that the correct thing is to split those 3 lines out as a
> separate commit that also inserts them into Perl_hv_ksplit, along with
> comments explaining what's going on, but I want to sleep on this idea.
>
The bucket chain randomization IMO is unnecessary with a good hash with a
128 bit seed like SipHash.

It is not even clear how much uncertainty it adds to the hashing.

consider we have something like this:

01234567
  A
  B

After a split we have 4 possibilities:

0123456789ABCDEF  # case 1, nothing moves, keyorder: AB
  A
  B

0123456789ABCDEF  # case 2, B moves, keyorder: AB
  A      B

0123456789ABCDEF  # case 3, A moves, keyorder: BA
  B      A

0123456789ABCDEF  # case 4, Both moved, keyorder: BA
         B
                  A

The bucket order randomization changes the last case so that it might also
produce:

0123456789ABCDEF  # case 4, Both moved, keyorder: AB
         A
                  B

The point is to make it harder to tell if a pair moved or not.  It is not
clear how well it works, with additional keys in the hash you can tease
apart the possibilities if you have an oracle hash function you can
consult. With this combined with the hash order traversal xor mask you need
more round trips to the oracle to tell if a pair of items collide. I added
this mostly because our hash function and seed allowed attacks on the seed,
which Siphash does not. I'd say that when SipHash is in use that logic
should be disabled. It should only be used in 32bit builds basically. With
a secure hash knowing that two keys collide shouldnt leak anything about
the seed.

cheers,
Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

hvds commented:
Apologies for going quiet on this; I did take a quick scan through the latest changes, and they look ok, but I haven't yet had time to go back through them more thoroughly. I hope to be able to make time for that in the next few days.
